File: macos/Sources/Chotki/Theme.swift
import SwiftUI
import AppKit

/// The palette from the design: deep charcoal ground, byzantine gold for today
/// and for what has been kept, muted violet for fasting, ochre for liturgy days.
///
/// Restraint is the point. Ornament reads as kitsch at 12pt, so the Orthodox
/// character comes from colour and typography rather than decoration.
enum Theme {
    static let ground = Color(red: 0.082, green: 0.086, blue: 0.110)
    static let panel = Color(red: 0.110, green: 0.118, blue: 0.149)
    static let line = Color(red: 0.180, green: 0.165, blue: 0.125)
    static let lineSoft = Color(red: 0.137, green: 0.145, blue: 0.173)

    static let gold = Color(red: 0.788, green: 0.635, blue: 0.153)
    static let goldDim = Color(red: 0.541, green: 0.447, blue: 0.125)
    static let parchment = Color(red: 0.910, green: 0.875, blue: 0.804)
    static let parchmentDim = Color(red: 0.847, green: 0.812, blue: 0.741)
    static let muted = Color(red: 0.541, green: 0.522, blue: 0.471)
    static let faint = Color(red: 0.353, green: 0.337, blue: 0.298)

    /// Fasting seasons. Never used to mark something as missed.
    static let violet = Color(red: 0.604, green: 0.561, blue: 0.769)
    /// Liturgy days. Note there is deliberately no "missed" colour anywhere in
    /// the palette — see the Tone section in design.md.
    static let ochre = Color(red: 0.651, green: 0.227, blue: 0.220)

    // MARK: the reading face

    /// The faces tried, in order, for anything meant to be read.
    ///
    /// **Iowan Old Style is not bundled and cannot be.** It is John Downer's,
    /// released through Bitstream and licensed to Apple; it ships with macOS
    /// and iOS but redistributing it in an app would be a licence violation.
    /// It also lives in *Supplemental*, which someone can disable in Font Book,
    /// so its presence is likely rather than certain — hence a chain.
    ///
    /// **Charter is the fallback on the merits, not merely by availability.**
    /// Matthew Carter, 1987, also originally Bitstream, drawn to the same brief
    /// as Iowan: large x-height, low stroke contrast, sturdy blunt serifs,
    /// engineered to hold up at text sizes. It is the nearest thing in the
    /// system to what Iowan does — and, unlike Iowan, it is freely
    /// redistributable, which is why it is the face the Android port will
    /// bundle.
    ///
    /// Family names rather than PostScript names, so `.bold()` and `.italic()`
    /// resolve to the real cuts instead of being synthesised.
    static let serifChain = ["Iowan Old Style", "Charter"]

    /// The first face in the chain this machine actually has, or nil.
    ///
    /// Resolved once, by asking AppKit. `Font.custom` falls back **silently**
    /// when a face is missing, which would make a broken chain invisible rather
    /// than obviously wrong — so the chain is walked rather than hoped at.
    static let readingFace: String? = serifChain.first { NSFont(name: $0, size: 12) != nil }

    /// The face for anything meant to be read: prayers, psalms, readings,
    /// glossary entries, reflections, the progress prose.
    ///
    /// Chrome keeps the system sans — the sidebar, the month grid's numerals,
    /// times, buttons and settings. Figures in a dense grid want to be tight
    /// and unambiguous, and AppKit controls do not take a custom face cleanly.
    ///
    /// `relativeTo:` is what keeps Dynamic Type working; `Font.custom(_:size:)`
    /// without it pins the size and quietly opts the app out.
    static func reading(
        _ size: CGFloat, relativeTo style: Font.TextStyle = .body
    ) -> Font {
        guard let readingFace else { return .system(size: size, design: .serif) }
        return .custom(readingFace, size: size, relativeTo: style)
    }

    static let popoverWidth: CGFloat = 400
    /// Fixed. The popover must not resize itself around whatever tab happens to
    /// be showing — a short tab would shrink it and leave every other tab
    /// scrolling inside a window that never grew back.
    static let popoverHeight: CGFloat = 560
}

extension View {
    /// A row that reads as tappable without shouting.
    func rowBackground(_ active: Bool = false) -> some View {
        background(active ? Theme.panel : Color.clear)
            .clipShape(RoundedRectangle(cornerRadius: 5))
    }
}

File: android/app/src/main/kotlin/org/chotki/app/ui/Theme.kt
package org.chotki.app.ui

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontFamily

/**
 * The same palette as the macOS app, to the value.
 *
 * Taken from `Theme.swift` rather than approximated: the two are one
 * application, and a person who keeps their rule on both should not feel they
 * have moved house. Dark throughout — the app is opened first thing in the
 * morning and last thing at night, and a white screen at either hour is unkind.
 */
object Chotki {
    val ground = Color(0xFF15161C)
    val panel = Color(0xFF1C1E26)
    val line = Color(0xFF2E2A20)
    val lineSoft = Color(0xFF23252C)

    val gold = Color(0xFFC9A227)
    val goldDim = Color(0xFF8A7220)
    val parchment = Color(0xFFE8DFCD)
    val parchmentDim = Color(0xFFD8CFBD)
    val muted = Color(0xFF8A8578)
    val faint = Color(0xFF5A564C)

    /**
     * The face for anything meant to be read: prayers, psalms, the day's
     * readings, the fathers, glossary entries, reflections.
     *
     * **Android's own serif, not a bundled one.** macOS and iOS read in Iowan
     * Old Style, which is John Downer's, licensed to Apple, and may not be
     * redistributed — so it cannot come here whatever the bundle size. The rule
     * agreed for the ports is that each platform picks the **native face
     * closest to what macOS chose**, and on Android that is the platform serif:
     * present on every device, no licence to check, no megabyte to ship.
     *
     * Charter — the second face in the Apple chain — is freely redistributable
     * and could be bundled if the platform serif is ever judged too far off.
     * That is a decision to take deliberately with the font in front of you,
     * not one to make by default.
     *
     * Chrome keeps the platform sans: the month grid's figures want to be tight
     * and unambiguous, and Roboto is what makes the app look like it belongs on
     * the device.
     */
    val reading = FontFamily.Serif

    val violet = Color(0xFF9A8FC4)
    val ochre = Color(0xFFA63A38)
}

@Composable
fun ChotkiTheme(content: @Composable () -> Unit) {
    // Dark either way. `isSystemInDarkTheme` is read so the intent is explicit
    // rather than accidental: this is not a light theme that happens to be dark.
    @Suppress("UNUSED_EXPRESSION") isSystemInDarkTheme()
    MaterialTheme(
        colorScheme = darkColorScheme(
            primary = Chotki.gold,
            onPrimary = Chotki.ground,
            background = Chotki.ground,
            onBackground = Chotki.parchment,
            surface = Chotki.panel,
            onSurface = Chotki.parchment,
            error = Chotki.ochre,
        ),
        content = content,
    )
}
